### ISSUE TITLE
Use utf8mb4 instead of utf8

### ISSUE THREAD

**Issue by Hazbelll** [sbpp/sourcebans-pp#246]
## Expected Behavior

From briefly looking through the [sourcecode,](https://github.com/sbpp/sourcebans-pp/search?utf8=%E2%9C%93&q=utf8) SourceBans++ currently uses `utf8` for its Charset, Collation, etc. Contrary to the name, `utf8` in MySQL does **NOT** fully support UTF-8 as it only supports a maximum of 3 bytes per character and only BMP characters (As briefly described [on the MySQL Dev Website](https://dev.mysql.com/doc/refman/5.5/en/charset-unicode-utf8mb4.html)). `utf8mb4` supports the entirety of UTF-8, thus characters such as emojis will not cause odd behaviour and/or the data doesn't need to be altered to remove such issues.

## Current Behavior

Characters that are non-BMP/contain 4 bytes will not be properly stored in the SourceBans++ database. A typical scenario where such data may be submitted would be a User being banned from the server who's Steam name contains emojis (Example: 🐌 **I_AM_SNAIL** 🐌). Some other examples can be found [on this blog](https://mathiasbynens.be/notes/mysql-utf8mb4).

## Possible Solution

Use `utf8mb4` instead of `utf8` throughout SourceBans++. However, the minimum supported version of MySQL by SourceBans++ will need to move up from **5.0** to **5.5.3**. It might be difficult/awkward to automatically changeover existing sourceban databases prior to such change without manual intervention by the user.

## Steps to Reproduce (for bugs)

1. Add emojis and/or other non-BMP/4-byte UTF-8 Characters into your Steam username.
2. Ban yourself via Sourcebans.
3. Check the data recorded into the database.

## Context

Mostly revolves around submitting the usernames of banned players who might contain UTF-8 characters that are non-BMP/4 bytes in size (such as emojis). Other scenarios might exist, but this is the most obvious one as users are free to set their usernames to whatever they desire.

## Your Environment

* Version used: https://github.com/sbpp/sourcebans-pp/commit/36ada1d72c64b6bde65578a71f97ab1b8df6b382
* PHP and MySQL version: PHP v7.0 x64 | MySQL 5.7 x64
* Operating System and version: Microsoft Windows Server 2012 R2 Datacenter x64
* Link to your project: N/A

*My apologies if anything above is inaccurate/incomprehensible - This isn't my area of expertise, but I thought I'd raise it just in case.*

**Comment by Groruk** [comment]
Hey, thank you for sharing this insight. Since I am currently getting all of this database mess fixed with switching from ADODB to PDO, I will include this for version 1.6 (which should hopefully be ready shortly). Every MySQL server running 5.5.3 or higher will be upgrading to utf8mb4, and everyone below this release has to rely on utf8 (some web hosters still only offer MySQL 5.1 ... what a shame).

---

> My apologies if anything above is inaccurate/incomprehensible - This isn't my area of expertise, but I thought I'd raise it just in case.

Have you seen other bug reports/suggestions here? You did an excellent job of detailing everything, keep up the good work.

**Comment by Hazbelll** [comment]
Sorry for the late reply, I had to pop out after I finished writing this up.

Thanks for the incredibly fast response! It's great to hear it's all being overhauled and this dealt as part of such. I look forward to 1.6, be sure to ping me when it's available! It is indeed a shame about those hosts that don't offer the latest versions. Hopefully this'll nudge them a little to upgrade. Every little helps.

> Have you seen other bug reports/suggestions here? You did an excellent job of detailing everything, keep up the good work.

No problem, thank you for the kind words. I'm happy to help anyway I can. Really appreciate the work you and others put into this project!
